In [1]:
import pandas as pd
import numpy as np
import datetime as dt
import math
#from datetime import date, datetime

In [2]:
# import and read train csv file.
weather_csv='./dataset/weather.csv'
weather=pd.read_csv(weather_csv)
# Set option to view all columns
pd.set_option('display.max_columns',50)
weather.head(20)

Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0         1  2007-05-01    83    50   67     14        51      56    0    2   
1         2  2007-05-01    84    52   68      M        51      57    0    3   
2         1  2007-05-02    59    42   51     -3        42      47   14    0   
3         2  2007-05-02    60    43   52      M        42      47   13    0   
4         1  2007-05-03    66    46   56      2        40      48    9    0   
5         2  2007-05-03    67    48   58      M        40      50    7    0   
6         1  2007-05-04    66    49   58      4        41      50    7    0   
7         2  2007-05-04    78    51    M      M        42      50    M    M   
8         1  2007-05-05    66    53   60      5        38      49    5    0   
9         2  2007-05-05    66    54   60      M        39      50    5    0   
10        1  2007-05-06    68    49   59      4        30      46    6    0   
11        2  2007-05-06    68    52   60      M        30      46    5    0   
12        1  2007-05-07    83    47   65     10        41      54    0    0   
13        2  2007-05-07    84    50   67      M        39      53    0    2   
14        1  2007-05-08    82    54   68     12        58      62    0    3   
15        2  2007-05-08    80    60   70      M        57      63    0    5   
16        1  2007-05-09    77    61   69     13        59      63    0    4   
17        2  2007-05-09    76    63   70      M        60      63    0    5   
18        1  2007-05-10    84    56   70     14        52      60    0    5   
19        2  2007-05-10    83    59   71      M        52      61    0    6   

   Sunrise Sunset CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
0     0448   1849             0      M      0.0        0.00       29.10   
1        -      -             M      M        M        0.00       29.18   
2     0447   1850      BR     0      M      0.0        0.00       29.38   
3        -      -   BR HZ     M      M        M        0.00       29.44   
4     0446   1851             0      M      0.0        0.00       29.39   
5        -      -      HZ     M      M        M        0.00       29.46   
6     0444   1852      RA     0      M      0.0           T       29.31   
7        -      -             M      M        M        0.00       29.36   
8     0443   1853             0      M      0.0           T       29.40   
9        -      -             M      M        M           T       29.46   
10    0442   1855             0      M      0.0        0.00       29.57   
11       -      -             M      M        M        0.00       29.62   
12    0441   1856      RA     0      M      0.0           T       29.38   
13       -      -             M      M        M        0.00       29.44   
14    0439   1857      BR     0      M      0.0        0.00       29.29   
15       -      -      HZ     M      M        M           T       29.36   
16    0438   1858   BR HZ     0      M      0.0        0.13       29.21   
17       -      -   BR HZ     M      M        M        0.02       29.28   
18    0437   1859      BR     0      M      0.0        0.00       29.20   
19       -      -   BR HZ     M      M        M        0.00       29.26   

   SeaLevel  ResultSpeed  ResultDir AvgSpeed  
0     29.82          1.7         27      9.2  
1     29.82          2.7         25      9.6  
2     30.09         13.0          4     13.4  
3     30.08         13.3          2     13.4  
4     30.12         11.7          7     11.9  
5     30.12         12.9          6     13.2  
6     30.05         10.4          8     10.8  
7     30.04         10.1          7     10.4  
8     30.10         11.7          7     12.0  
9     30.09         11.2          7     11.5  
10    30.29         14.4         11     15.0  
11    30.28         13.8         10     14.5  
12    30.12          8.6         18     10.5  
13    30.12          8.5         17      9.9  
14    30.03          2.7         11      5.8  
15    30.02     

In [ ]:
# 
#weather['Year']=pd.DatetimeIndex(weather['Date']).year
#weather['Month']=pd.DatetimeIndex(weather['Date']).month
#weather['Day']=pd.DatetimeIndex(weather['Date']).day
#weather['YYMM']=weather['Year'].astype(str)+weather['Month'].astype(str)
#weather.head()

In [3]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 22 columns):
Station        2944 non-null int64
Date           2944 non-null object
Tmax           2944 non-null int64
Tmin           2944 non-null int64
Tavg           2944 non-null object
Depart         2944 non-null object
DewPoint       2944 non-null int64
WetBulb        2944 non-null object
Heat           2944 non-null object
Cool           2944 non-null object
Sunrise        2944 non-null object
Sunset         2944 non-null object
CodeSum        2944 non-null object
Depth          2944 non-null object
Water1         2944 non-null object
SnowFall       2944 non-null object
PrecipTotal    2944 non-null object
StnPressure    2944 non-null object
SeaLevel       2944 non-null object
ResultSpeed    2944 non-null float64
ResultDir      2944 non-null int64
AvgSpeed       2944 non-null object
dtypes: float64(1), int64(5), object(16)
memory usage: 506.1+ KB


#### columns with blank ' -'  and  M 
- Sunrise - 1472 (stn 2)--> populated from Stn 1
- Sunset- 1472 (stn 2)--> populated from Stn 1
- Depart-1472 (stn 2)--> populated from Stn 1
- Depth- 1472 (stn 2)--> populated from Stn 1
- SnowFall-1472 (stn 2) --> populated from Stn 1


#### No date pattern spotted (to be updated with the Month' Median)
- WetBulb-4  (stn 1 & 2) 26/6/09, 10/08/13, 11/08/13, 12/08/13 (2)
- StnPressure- 4 (stn 1 & 2) 13/06/07, 26/06/09, 10/08/13, 10/08/13
- SeaLevel- 9 (stn 1 & 2)

#### Get from stn 1
- PrecipTotal- 2 (stn 2)
- AvgSpeed- 3 (stn 2)

#### same date pattern spotted 
- Heat- 11    (stn 2)--> populated from Stn 1
- Cool- 11 (stn 2)   --> populated from Stn 1
- Tavg-11     (stn 2,--> can be calculated based on Tmax, Tmin)


#### remain
- Water1- 2944 (stn 1 & 2)- No action taken

#### columns with empty string
- CodeSum- 1609 (stn 1 & 2) ---> filled with 'No Observation'

1. Update `Tavg` Missing Data by taking average of `Tmax` and `Tmin` and change `Tavg` datatype to float.

In [4]:
weather[weather['Tavg']=='M']
# index row 7, 505, 675, 1637, 2067, 2211, 2501, 2511, 2525, 2579,2811

Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
7           2  2007-05-04    78    51    M      M        42      50    M    M   
505         2  2008-07-08    86    46    M      M        68      71    M    M   
675         2  2008-10-01    62    46    M      M        41      47    M    M   
1637        2  2011-07-22   100    71    M      M        70      74    M    M   
2067        2  2012-08-22    84    72    M      M        51      61    M    M   
2211        2  2013-05-02    71    42    M      M        39      45    M    M   
2501        2  2013-09-24    91    52    M      M        48      54    M    M   
2511        2  2013-09-29    84    53    M      M        48      54    M    M   
2525        2  2013-10-06    76    48    M      M        44      50    M    M   
2579        2  2014-05-02    80    47    M      M        43      47    M    M   
2811        2  2014-08-26    86    49    M      M        68      71    M    M   

     Sunrise Sunset     CodeSum Depth Water1 SnowFall PrecipTotal StnPressure  \
7          -      -                 M      M        M        0.00       29.36   
505        -      -       TS RA     M      M        M        0.28       29.16   
675        -      -                 M      M        M        0.00       29.30   
1637       -      -  TS TSRA BR     M      M        M        0.14       29.23   
2067       -      -                 M      M        M        0.00       29.39   
2211       -      -                 M      M        M        0.00       29.51   
2501       -      -                 M      M        M        0.00       29.33   
2511       -      -       RA BR     M      M        M        0.22       29.36   
2525       -      -    RA DZ BR     M      M        M        0.06       29.10   
2579       -      -          RA     M      M        M        0.04       29.10   
2811       -      -                 M      M        M           T       29.41   

     SeaLevel  ResultSpeed  ResultDir AvgSpeed  
7       30.04         10.1          7     10.4  
505     29.80          7.4         24      8.3  
675     29.96         10.9         33     11.0  
1637    29.86          3.8         10      8.2  
2067        M          4.7         19        M  
2211    30.17         15.8          2     16.1  
2501    30.00          5.8          9      7.7  
2511    30.01          6.3         36      7.8  
2525    29.76         10.1         25     10.6  
2579    29.79         10.7         23     11.9  
2811    30.04          1.3         15      5.8

In [5]:
weather.loc[weather['Tavg'] == 'M', 'Tavg'] =round(((weather['Tmax']+weather['Tmin'])/2))

In [10]:
# check result after update. Tavg was not 'Round-Up', just roudn to nearest int
weather.iloc[[ 7, 505, 675, 1637, 2067, 2211, 2501, 2511, 2525, 2579,2811],0:5]

Station        Date  Tmax  Tmin Tavg
7           2  2007-05-04    78    51   64
505         2  2008-07-08    86    46   66
675         2  2008-10-01    62    46   54
1637        2  2011-07-22   100    71   86
2067        2  2012-08-22    84    72   78
2211        2  2013-05-02    71    42   56
2501        2  2013-09-24    91    52   72
2511        2  2013-09-29    84    53   68
2525        2  2013-10-06    76    48   62
2579        2  2014-05-02    80    47   64
2811        2  2014-08-26    86    49   68

In [11]:
weather['Tavg']=pd.to_numeric(weather['Tavg'])
print(weather['Tavg'].dtypes)

float64


2. Create a df without date duplication.

In [12]:
station_1=weather[weather['Station']==1]
station_2=weather[weather['Station']==2]
combined=station_1.merge(station_2,on='Date',how='inner',suffixes=('_1','_2'))
print(combined.shape)
combined.head(2)

(1472, 43)


Station_1        Date  Tmax_1  Tmin_1  Tavg_1 Depart_1  DewPoint_1  \
0          1  2007-05-01      83      50    67.0       14          51   
1          1  2007-05-02      59      42    51.0       -3          42   

  WetBulb_1 Heat_1 Cool_1 Sunrise_1 Sunset_1 CodeSum_1 Depth_1 Water1_1  \
0        56      0      2      0448     1849                 0        M   
1        47     14      0      0447     1850        BR       0        M   

  SnowFall_1 PrecipTotal_1 StnPressure_1 SeaLevel_1  ResultSpeed_1  \
0        0.0          0.00         29.10      29.82            1.7   
1        0.0          0.00         29.38      30.09           13.0   

   ResultDir_1 AvgSpeed_1  Station_2  Tmax_2  Tmin_2  Tavg_2 Depart_2  \
0           27        9.2          2      84      52    68.0        M   
1            4       13.4          2      60      43    52.0        M   

   DewPoint_2 WetBulb_2 Heat_2 Cool_2 Sunrise_2 Sunset_2 CodeSum_2 Depth_2  \
0          51        57      0      3         -        -                 M   
1          42        47     13      0         -        -     BR HZ       M   

  Water1_2 SnowFall_2 PrecipTotal_2 StnPressure_2 SeaLevel_2  ResultSpeed_2  \
0        M          M          0.00         29.18      29.82            2.7   
1        M          M          0.00         29.44      30.08           13.3   

   ResultDir_2 AvgSpeed_2  
0           25        9.6  
1            2       13.4

3. Update `Heat`,`Cool`Missing Data in Stn 2 with information from Stn 1 & change data type to int

In [13]:
check_heat_cool=combined[(combined['Heat_2']=='M')|(combined['Cool_2']=='M')]
check_heat_cool_miss=check_heat_cool[['Station_1','Station_2','Heat_1','Heat_2','Cool_1','Cool_2']]
check_heat_cool_miss
#iloc=[3,252,337,818,1033,1105,1250,1255,1262,1289,1405]

Station_1  Station_2 Heat_1 Heat_2 Cool_1 Cool_2
3             1          2      7      M      0      M
252           1          2      0      M     13      M
337           1          2     11      M      0      M
818           1          2      0      M     14      M
1033          1          2      0      M      5      M
1105          1          2     16      M      0      M
1250          1          2      5      M      0      M
1255          1          2      5      M      0      M
1262          1          2     10      M      0      M
1289          1          2     15      M      0      M
1405          1          2      0      M     13      M

In [14]:
combined.loc[combined['Heat_2']=='M','Heat_2']=combined['Heat_1']
combined.loc[combined['Cool_2']=='M','Cool_2']=combined['Cool_1']

In [15]:
# Check after amendment
combined.iloc[[ 3,252,337,818,1033,1105,1250,1255,1262,1289,1405],[0,8,9,22,29,30]]

Station_1 Heat_1 Cool_1  Station_2 Heat_2 Cool_2
3             1      7      0          2      7      0
252           1      0     13          2      0     13
337           1     11      0          2     11      0
818           1      0     14          2      0     14
1033          1      0      5          2      0      5
1105          1     16      0          2     16      0
1250          1      5      0          2      5      0
1255          1      5      0          2      5      0
1262          1     10      0          2     10      0
1289          1     15      0          2     15      0
1405          1      0     13          2      0     13

In [16]:
# change datatype for Heat_1','Heat_2','Cool_1','Cool_2'

combined[['Heat_1','Heat_2','Cool_1','Cool_2']]=combined[['Heat_1','Heat_2','Cool_1','Cool_2']].apply(pd.to_numeric)
print(combined['Heat_1'].dtypes)
print(combined['Heat_2'].dtypes)
print(combined['Cool_1'].dtypes)
print(combined['Cool_2'].dtypes)


int64
int64
int64
int64


4. Update `PrecipTotal`,`AvgSpeed`Missing Data in Stn 2 with information from Stn 1. No changes on datatype as Precip consist 'T'

In [17]:
check_precip_speed=combined[(combined['PrecipTotal_2']=='M')|(combined['AvgSpeed_2']=='M')]
check_precip_speed_miss=check_precip_speed[['Station_1','Station_2','PrecipTotal_1','AvgSpeed_1','PrecipTotal_2','AvgSpeed_2']]
check_precip_speed_miss
# iloc=[43,58,59,872,1033]

Station_1  Station_2 PrecipTotal_1 AvgSpeed_1 PrecipTotal_2 AvgSpeed_2
43            1          2          0.00        8.6          0.00          M
58            1          2          0.00       12.5             M       13.3
59            1          2          0.00        8.3             M        8.2
872           1          2          0.08        7.3             T          M
1033          1          2          0.00        5.8          0.00          M

In [18]:
combined.loc[combined['PrecipTotal_2']=='M','PrecipTotal_2']=combined['PrecipTotal_1']
combined.loc[combined['AvgSpeed_2']=='M','AvgSpeed_2']=combined['AvgSpeed_1']

In [19]:
# Check after amendment
combined.iloc[[ 43,58,59,872,1033],[0,16,21,22,37,42]]

Station_1 PrecipTotal_1 AvgSpeed_1  Station_2 PrecipTotal_2 AvgSpeed_2
43            1          0.00        8.6          2          0.00        8.6
58            1          0.00       12.5          2          0.00       13.3
59            1          0.00        8.3          2          0.00        8.2
872           1          0.08        7.3          2             T        7.3
1033          1          0.00        5.8          2          0.00        5.8

In [20]:
# change datatype for AvgSpeed_1','AvgSpeed_2'

combined[['AvgSpeed_1','AvgSpeed_2']]=combined[['AvgSpeed_1','AvgSpeed_2']].apply(pd.to_numeric)
print(combined['PrecipTotal_1'].dtypes)
print(combined['PrecipTotal_2'].dtypes)
print(combined['AvgSpeed_1'].dtypes)
print(combined['AvgSpeed_2'].dtypes)


object
object
float64
float64


5. Update `WetBulb`,`StnPressure`,`SeaLevel` Missing Data with respective columns' Median.

In [22]:
check_WetBulb=combined[(combined['WetBulb_2']=='M')|(combined['WetBulb_1']=='M')|(combined['StnPressure_1']=='M')|(combined['StnPressure_2']=='M')|(combined['SeaLevel_1']=='M')|(combined['SeaLevel_2']=='M')]
check_WetBulb_miss=check_WetBulb[['Station_1','Station_2','WetBulb_2','WetBulb_1','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2']]
check_WetBulb_miss
# iloc=[43,416,424,497,866,872,878,1033,1045,1205,1206,1207,1371,1,2,3,1201,1202,1203]

Station_1  Station_2 WetBulb_2 WetBulb_1 StnPressure_1 StnPressure_2  \
43            1          2        62        62         29.36             M   
416           1          2        67        67         29.08         29.15   
424           1          2        67         M             M         29.20   
497           1          2        63        62         29.39         29.44   
866           1          2        59        59         29.34         29.36   
872           1          2        51        49         29.39         29.47   
878           1          2        58        58         29.26         29.33   
1033          1          2        61        60         29.35         29.39   
1045          1          2        73        73         29.17         29.23   
1205          1          2        63         M             M             M   
1206          1          2        64         M         29.35         29.42   
1207          1          2         M        68         29.21         29.27   
1371          1          2        61        58         29.41         29.47   

     SeaLevel_1 SeaLevel_2  
43        30.09          M  
416           M      29.79  
424       29.85      29.83  
497           M      30.09  
866           M      30.03  
872       30.09          M  
878           M      29.99  
1033      30.07          M  
1045          M      29.87  
1205      30.08      30.07  
1206      30.07      30.06  
1207      29.93      29.92  
1371      30.12          M

In [23]:
len(check_WetBulb_miss)

13

In [24]:
# change 'M' to NaN to calculte column median
combined[['WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2']]=combined[['WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2']].replace('M',np.nan)

In [27]:
combined.iloc[[43,416,424,497,866,872,878,1033,1045,1205,1206,1207,1371],[7,28,17,38,18,39]]

WetBulb_1 WetBulb_2 StnPressure_1 StnPressure_2 SeaLevel_1 SeaLevel_2
43          62        62         29.36           NaN      30.09        NaN
416         67        67         29.08         29.15        NaN      29.79
424        NaN        67           NaN         29.20      29.85      29.83
497         62        63         29.39         29.44        NaN      30.09
866         59        59         29.34         29.36        NaN      30.03
872         49        51         29.39         29.47      30.09        NaN
878         58        58         29.26         29.33        NaN      29.99
1033        60        61         29.35         29.39      30.07        NaN
1045        73        73         29.17         29.23        NaN      29.87
1205       NaN        63           NaN           NaN      30.08      30.07
1206       NaN        64         29.35         29.42      30.07      30.06
1207        68       NaN         29.21         29.27      29.93      29.92
1371        58        61         29.41         29.47      30.12        NaN

In [28]:
print(combined['WetBulb_1'].median())
print(combined['WetBulb_2'].median())
print(combined['StnPressure_1'].median())
print(combined['StnPressure_2'].median())
print(combined['SeaLevel_1'].median())
print(combined['SeaLevel_2'].median())

60.0
61.0
29.26
29.32
29.98
29.96


In [29]:
combined.loc[combined['WetBulb_1'].isnull(),'WetBulb_1']=combined['WetBulb_1'].median()
combined.loc[combined['WetBulb_2'].isnull(),'WetBulb_2']=combined['WetBulb_2'].median()
combined.loc[combined['StnPressure_1'].isnull(),'StnPressure_1']=combined['StnPressure_1'].median()
combined.loc[combined['StnPressure_2'].isnull(),'StnPressure_2']=combined['StnPressure_2'].median()
combined.loc[combined['SeaLevel_1'].isnull(),'SeaLevel_1']=combined['SeaLevel_1'].median()
combined.loc[combined['SeaLevel_2'].isnull(),'SeaLevel_2']=combined['SeaLevel_2'].median()

In [30]:
# Check after amendment
combined.iloc[[43,416,424,497,866,872,878,1033,1045,1205,1206,1207,1371],[7,28,17,38,18,39]]

WetBulb_1 WetBulb_2 StnPressure_1 StnPressure_2 SeaLevel_1 SeaLevel_2
43          62        62         29.36         29.32      30.09      29.96
416         67        67         29.08         29.15      29.98      29.79
424         60        67         29.26         29.20      29.85      29.83
497         62        63         29.39         29.44      29.98      30.09
866         59        59         29.34         29.36      29.98      30.03
872         49        51         29.39         29.47      30.09      29.96
878         58        58         29.26         29.33      29.98      29.99
1033        60        61         29.35         29.39      30.07      29.96
1045        73        73         29.17         29.23      29.98      29.87
1205        60        63         29.26         29.32      30.08      30.07
1206        60        64         29.35         29.42      30.07      30.06
1207        68        61         29.21         29.27      29.93      29.92
1371        58        61         29.41         29.47      30.12      29.96

In [31]:
# change datatype for 'WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2'

combined[['WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2']]=combined[['WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2']].apply(pd.to_numeric)
print(combined['WetBulb_1'].dtypes)
print(combined['WetBulb_2'].dtypes)
print(combined['StnPressure_1'].dtypes)
print(combined['StnPressure_2'].dtypes)
print(combined['SeaLevel_1'].dtypes)
print(combined['SeaLevel_2'].dtypes)

float64
float64
float64
float64
float64
float64


6. Update `Depart`,`Depth`,`Sunrise`,`Sunset`,`SnowFall`,` Missing Data in all 1472 Station 2 observation are updated with Station 1 information as only Station 1 has the figure recorded. Changed datatype for Depart,Depth, Sunrise, Sunset。 Except for SnowFall as it contains 'T'


In [32]:
# Depart,Depth, Sunrise, Sunset, SnowFall, map info from station 1 to station 2

combined['Depart_2']=combined['Depart_1']
combined['Depth_2']=combined['Depth_1']
combined['Sunrise_2']=combined['Sunrise_1']
combined['Sunset_2']=combined['Sunset_1']
combined['SnowFall_2']=combined['SnowFall_1']
combined[['Depart_1','Depart_2','Depth_1','Depth_2','Sunrise_1','Sunrise_2','Sunset_1','Sunset_2','SnowFall_1','SnowFall_2',]]


Depart_1 Depart_2 Depth_1 Depth_2 Sunrise_1 Sunrise_2 Sunset_1 Sunset_2  \
0          14       14       0       0      0448      0448     1849     1849   
1          -3       -3       0       0      0447      0447     1850     1850   
2           2        2       0       0      0446      0446     1851     1851   
3           4        4       0       0      0444      0444     1852     1852   
4           5        5       0       0      0443      0443     1853     1853   
5           4        4       0       0      0442      0442     1855     1855   
6          10       10       0       0      0441      0441     1856     1856   
7          12       12       0       0      0439      0439     1857     1857   
8          13       13       0       0      0438      0438     1858     1858   
9          14       14       0       0      0437      0437     1859     1859   
10          4        4       0       0      0436      0436     1860     1860   
11         -2       -2       0       0      0435      0435     1901     1901   
12         -2       -2       0       0      0434      0434     1902     1902   
13         15       15       0       0      0433      0433     1903     1903   
14         11       11       0       0      0432      0432     1904     1904   
15         -4       -4       0       0      0431      0431     1905     1905   
16         -6       -6       0       0      0430      0430     1906     1906   
17         -3       -3       0       0      0429      0429     1907     1907   
18          8        8       0       0      0428      0428     1908     1908   
19         -5       -5       0       0      0427      0427     1909     1909   
20          1        1       0       0      0426      0426     1910     1910   
21         10       10       0       0      0425      0425     1911     1911   
22         14       14       0       0      0425      0425     1912     1912   
23         14       14       0       0      0424      0424     1913     1913   
24          1        1       0       0      0423      0423     1914     1914   
25         -5       -5       0       0      0422      0422     1915     1915   
26          4        4       0       0      0422      0422     1916     1916   
27          2        2       0       0      0421      0421     1916     1916   
28         10       10       0       0      0421      0421     1917     1917   
29         13       13       0       0      0420      0420     1918     1918   
...       ...      ...     ...     ...       ...       ...      ...      ...   
1442       10       10       0       0      0549      0549     1732     1732   
1443       -3       -3       0       0      0550      0550     1730     1730   
1444      -16      -16       0       0      0552      0552     1729     1729   
1445      -11      -11       0       0      0553      0553     1727     1727   
1446       -4       -4       0       0      0554      0554     1725     1725   
1447        1        1       0       0      0555      0555     1724     1724   
1448       -1       -1       0       0      0556      0556     1722     1722   
1449       -2       -2       0       0      0557      0557     1720     1720   
1450       -6       -6       0       0      0558      0558     1719     1719   
1451       -6       -6       0       0      0559      0559     1717     1717   
1452       -3       -3       0       0      0600      0600     1715     1715   
1453        9        9       0       0      0602      0602     1714     1714   
1454        6        6       0       0      0603      0603     1712     1712   
1455        0        0       0       0      0604      0604     1711     1711   
1456        1        1       0       0      0605      0605     1709     1709   
1457        3        3       0       0      0606      0606     1707     1707   
1458       -8       -8       0       0      0607      0607     1706     1706   
1459       -7       -7       0       0      0608      0608     1704     1704   
1460 

In [33]:
# change datatype for 'WetBulb_1','WetBulb_2','StnPressure_1','StnPressure_2','SeaLevel_1','SeaLevel_2'

combined[['Depart_1','Depart_2','Depth_1','Depth_2','Sunrise_1','Sunrise_2','Sunset_1','Sunset_2']]=combined[['Depart_1','Depart_2','Depth_1','Depth_2','Sunrise_1','Sunrise_2','Sunset_1','Sunset_2']].apply(pd.to_numeric)
print(combined['Depart_1'].dtypes)
print(combined['Depart_2'].dtypes)
print(combined['Depth_1'].dtypes)
print(combined['Depth_2'].dtypes)
print(combined['Sunrise_1'].dtypes)
print(combined['Sunrise_2'].dtypes)
print(combined['Sunset_1'].dtypes)
print(combined['Sunset_2'].dtypes)


int64
int64
int64
int64
int64
int64
int64
int64


7. Update 1609 `CodeSum` Missing Data in both Station 1 & Station 2 with value 'No Observation'.


In [34]:
combined[['CodeSum_1','CodeSum_2','Date']]

CodeSum_1   CodeSum_2        Date
0                                2007-05-01
1                BR       BR HZ  2007-05-02
2                            HZ  2007-05-03
3                RA              2007-05-04
4                                2007-05-05
5                                2007-05-06
6                RA              2007-05-07
7                BR          HZ  2007-05-08
8             BR HZ       BR HZ  2007-05-09
9                BR       BR HZ  2007-05-10
10                               2007-05-11
11                               2007-05-12
12                               2007-05-13
13                               2007-05-14
14            RA BR  TSRA RA BR  2007-05-15
15            RA BR       RA BR  2007-05-16
16                               2007-05-17
17                               2007-05-18
18                               2007-05-19
19          RA VCTS     TSRA RA  2007-05-20
20                               2007-05-21
21                               2007-05-22
22                               2007-05-23
23            RA HZ       RA HZ  2007-05-24
24                        RA BR  2007-05-25
25    TSRA RA BR HZ  TSRA RA BR  2007-05-26
26            RA BR       RA BR  2007-05-27
27               RA              2007-05-28
28            BR HZ       BR HZ  2007-05-29
29               HZ          HZ  2007-05-30
...             ...         ...         ...
1442     TSRA RA BR  TSRA RA BR  2014-10-02
1443             RA       RA BR  2014-10-03
1444       RA DZ BR       RA DZ  2014-10-04
1445             RA          RA  2014-10-05
1446                             2014-10-06
1447                             2014-10-07
1448                             2014-10-08
1449                             2014-10-09
1450                             2014-10-10
1451                             2014-10-11
1452       RA DZ BR    RA DZ BR  2014-10-12
1453       RA DZ BR    RA BR HZ  2014-10-13
1454       RA DZ BR    RA DZ BR  2014-10-14
1455       RA DZ BR    RA DZ BR  2014-10-15
1456          DZ BR       DZ BR  2014-10-16
1457                             2014-10-17
1458          RA DZ          RA  2014-10-18
1459                             2014-10-19
1460                             2014-10-20
1461                             2014-10-21
1462                             2014-10-22
1463             BR          BR  2014-10-23
1464                             2014-10-24
1465             BR       BR HZ  2014-10-25
1466                             2014-10-26
1467                         RA  2014-10-27
1468                         RA  2014-10-28
1469                             2014-10-29
1470                         RA  2014-10-30
1471          RA SN    RA SN BR  2014-10-31

[1472 rows x 3 columns]

In [35]:
check_codeSum=combined[(combined['CodeSum_2']==' ')|(combined['CodeSum_1']==' ')]
check_codeSum_miss=check_codeSum[['Station_1','Station_2','CodeSum_1','CodeSum_2']]
check_codeSum_miss


Station_1  Station_2 CodeSum_1 CodeSum_2
0             1          2                    
2             1          2                  HZ
3             1          2        RA          
4             1          2                    
5             1          2                    
6             1          2        RA          
10            1          2                    
11            1          2                    
12            1          2                    
13            1          2                    
16            1          2                    
17            1          2                    
18            1          2                    
20            1          2                    
21            1          2                    
22            1          2                    
24            1          2               RA BR
27            1          2        RA          
37            1          2                    
38            1          2                    
39            1          2                    
40            1          2                    
41            1          2                    
42            1          2                    
43            1          2                    
44            1          2                    
45            1          2                    
50            1          2                    
58            1          2                    
59            1          2                    
...         ...        ...       ...       ...
1424          1          2                    
1426          1          2                    
1427          1          2                    
1428          1          2   BCFG BR          
1429          1          2        BR          
1431          1          2        RA          
1432          1          2                    
1433          1          2                    
1434          1          2                    
1435          1          2                    
1436          1          2        BR          
1437          1          2                    
1441          1          2                  RA
1446          1          2                    
1447          1          2                    
1448          1          2                    
1449          1          2                    
1450          1          2                    
1451          1          2                    
1457          1          2                    
1459          1          2                    
1460          1          2                    
1461          1          2                    
1462          1          2                    
1464          1          2                    
1466          1          2                    
1467          1          2                  RA
1468          1          2                  RA
1469          1          2                    
1470          1          2                  RA

[892 rows x 4 columns]

In [36]:
combined['CodeSum_1'].replace(' ','No Observation',inplace=True)
combined['CodeSum_2'].replace(' ','No Observation',inplace=True)

In [38]:
# Check after amendment
combined.iloc[[ 2,3,4,5,6],[12,33]]

CodeSum_1       CodeSum_2
2  No Observation              HZ
3              RA  No Observation
4  No Observation  No Observation
5  No Observation  No Observation
6              RA  No Observation

In [39]:
combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1472 entries, 0 to 1471
Data columns (total 43 columns):
Station_1        1472 non-null int64
Date             1472 non-null object
Tmax_1           1472 non-null int64
Tmin_1           1472 non-null int64
Tavg_1           1472 non-null float64
Depart_1         1472 non-null int64
DewPoint_1       1472 non-null int64
WetBulb_1        1472 non-null float64
Heat_1           1472 non-null int64
Cool_1           1472 non-null int64
Sunrise_1        1472 non-null int64
Sunset_1         1472 non-null int64
CodeSum_1        1472 non-null object
Depth_1          1472 non-null int64
Water1_1         1472 non-null object
SnowFall_1       1472 non-null object
PrecipTotal_1    1472 non-null object
StnPressure_1    1472 non-null float64
SeaLevel_1       1472 non-null float64
ResultSpeed_1    1472 non-null float64
ResultDir_1      1472 non-null int64
AvgSpeed_1       1472 non-null float64
Station_2        1472 non-null int64
Tmax_2           1472 non

In [26]:
columns_list=combined.columns.to_list()
for c in enumerate (columns_list):
    print(c)

(0, 'Station_1')
(1, 'Date')
(2, 'Tmax_1')
(3, 'Tmin_1')
(4, 'Tavg_1')
(5, 'Depart_1')
(6, 'DewPoint_1')
(7, 'WetBulb_1')
(8, 'Heat_1')
(9, 'Cool_1')
(10, 'Sunrise_1')
(11, 'Sunset_1')
(12, 'CodeSum_1')
(13, 'Depth_1')
(14, 'Water1_1')
(15, 'SnowFall_1')
(16, 'PrecipTotal_1')
(17, 'StnPressure_1')
(18, 'SeaLevel_1')
(19, 'ResultSpeed_1')
(20, 'ResultDir_1')
(21, 'AvgSpeed_1')
(22, 'Station_2')
(23, 'Tmax_2')
(24, 'Tmin_2')
(25, 'Tavg_2')
(26, 'Depart_2')
(27, 'DewPoint_2')
(28, 'WetBulb_2')
(29, 'Heat_2')
(30, 'Cool_2')
(31, 'Sunrise_2')
(32, 'Sunset_2')
(33, 'CodeSum_2')
(34, 'Depth_2')
(35, 'Water1_2')
(36, 'SnowFall_2')
(37, 'PrecipTotal_2')
(38, 'StnPressure_2')
(39, 'SeaLevel_2')
(40, 'ResultSpeed_2')
(41, 'ResultDir_2')
(42, 'AvgSpeed_2')


In [41]:
list(combined.columns)

['Station_1',
 'Date',
 'Tmax_1',
 'Tmin_1',
 'Tavg_1',
 'Depart_1',
 'DewPoint_1',
 'WetBulb_1',
 'Heat_1',
 'Cool_1',
 'Sunrise_1',
 'Sunset_1',
 'CodeSum_1',
 'Depth_1',
 'Water1_1',
 'SnowFall_1',
 'PrecipTotal_1',
 'StnPressure_1',
 'SeaLevel_1',
 'ResultSpeed_1',
 'ResultDir_1',
 'AvgSpeed_1',
 'Station_2',
 'Tmax_2',
 'Tmin_2',
 'Tavg_2',
 'Depart_2',
 'DewPoint_2',
 'WetBulb_2',
 'Heat_2',
 'Cool_2',
 'Sunrise_2',
 'Sunset_2',
 'CodeSum_2',
 'Depth_2',
 'Water1_2',
 'SnowFall_2',
 'PrecipTotal_2',
 'StnPressure_2',
 'SeaLevel_2',
 'ResultSpeed_2',
 'ResultDir_2',
 'AvgSpeed_2']

In [59]:
# segregate Station_1
station1_df=combined.filter(regex='_1')
station1_df['Date']=combined['Date']
station1_df['DateStation']=station1_df['Date'].astype(str)+station1_df['Station_1'].astype(str)
print(station1_df.shape)
station1_df


(1472, 23)


C:\Users\User\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\User\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Station_1  Tmax_1  Tmin_1  Tavg_1  Depart_1  DewPoint_1  WetBulb_1  \
0             1      83      50    67.0        14          51       56.0   
1             1      59      42    51.0        -3          42       47.0   
2             1      66      46    56.0         2          40       48.0   
3             1      66      49    58.0         4          41       50.0   
4             1      66      53    60.0         5          38       49.0   
5             1      68      49    59.0         4          30       46.0   
6             1      83      47    65.0        10          41       54.0   
7             1      82      54    68.0        12          58       62.0   
8             1      77      61    69.0        13          59       63.0   
9             1      84      56    70.0        14          52       60.0   
10            1      70      51    61.0         4          42       51.0   
11            1      64      46    55.0        -2          36       46.0   
12            1      69      43    56.0        -2          33       46.0   
13            1      90      56    73.0        15          47       59.0   
14            1      80      57    69.0        11          56       61.0   
15            1      66      44    55.0        -4          43       48.0   
16            1      62      43    53.0        -6          32       43.0   
17            1      71      42    57.0        -3          30       46.0   
18            1      80      56    68.0         8          39       53.0   
19            1      69      43    56.0        -5          40       44.0   
20            1      80      44    62.0         1          50       56.0   
21            1      87      55    71.0        10          46       58.0   
22            1      89      60    75.0        14          49       61.0   
23            1      88      63    76.0        14          54       63.0   
24            1      70      56    63.0         1          44       53.0   
25            1      65      49    57.0        -5          55       56.0   
26            1      75      57    66.0         4          47       56.0   
27            1      79      50    65.0         2          47       56.0   
28            1      88      60    74.0        10          58       65.0   
29            1      87      66    77.0        13          58       65.0   
...         ...     ...     ...     ...       ...         ...        ...   
1442          1      72      63    68.0        10          61       63.0   
1443          1      65      43    54.0        -3          45       49.0   
1444          1      47      35    41.0       -16          34       39.0   
1445          1      55      37    46.0       -11          35       42.0   
1446          1      61      42    52.0        -4          38       45.0   
1447          1      66      48    57.0         1          43       50.0   
1448          1      65      44    55.0        -1          35       46.0   
1449          1      62      43    53.0        -2          34       44.0   
1450          1      58      39    49.0        -6          39       44.0   
1451          1      59      36    48.0        -6          34       42.0   
1452          1      63      38    51.0        -3          42       47.0   
1453          1      67      56    62.0         9          58       60.0   
1454          1      65      52    59.0         6          57       58.0   
1455          1      55      50    53.0         0          51       51.0   
1456          1      56      50    53.0         1          49       51.0   
1457          1      62      48    55.0         3          46       50.0   
1458          1      50      36    43.0        -8          38       42.0   
1459          1      54      33    44.0        -7          38       43.0   
1460          1      65      46    56.0         6          44       49.0   
1461          1      53      45    49.0        -1          39       44.0   
1462          1      56      35    46.0        -4     

In [61]:
#Remove suffix
station1_df.columns = [col.replace('_1', '') for col in station1_df.columns]
station1_df

Station  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  Cool  \
0           1    83    50  67.0      14        51     56.0     0     2   
1           1    59    42  51.0      -3        42     47.0    14     0   
2           1    66    46  56.0       2        40     48.0     9     0   
3           1    66    49  58.0       4        41     50.0     7     0   
4           1    66    53  60.0       5        38     49.0     5     0   
5           1    68    49  59.0       4        30     46.0     6     0   
6           1    83    47  65.0      10        41     54.0     0     0   
7           1    82    54  68.0      12        58     62.0     0     3   
8           1    77    61  69.0      13        59     63.0     0     4   
9           1    84    56  70.0      14        52     60.0     0     5   
10          1    70    51  61.0       4        42     51.0     4     0   
11          1    64    46  55.0      -2        36     46.0    10     0   
12          1    69    43  56.0      -2        33     46.0     9     0   
13          1    90    56  73.0      15        47     59.0     0     8   
14          1    80    57  69.0      11        56     61.0     0     4   
15          1    66    44  55.0      -4        43     48.0    10     0   
16          1    62    43  53.0      -6        32     43.0    12     0   
17          1    71    42  57.0      -3        30     46.0     8     0   
18          1    80    56  68.0       8        39     53.0     0     3   
19          1    69    43  56.0      -5        40     44.0     9     0   
20          1    80    44  62.0       1        50     56.0     3     0   
21          1    87    55  71.0      10        46     58.0     0     6   
22          1    89    60  75.0      14        49     61.0     0    10   
23          1    88    63  76.0      14        54     63.0     0    11   
24          1    70    56  63.0       1        44     53.0     2     0   
25          1    65    49  57.0      -5        55     56.0     8     0   
26          1    75    57  66.0       4        47     56.0     0     1   
27          1    79    50  65.0       2        47     56.0     0     0   
28          1    88    60  74.0      10        58     65.0     0     9   
29          1    87    66  77.0      13        58     65.0     0    12   
...       ...   ...   ...   ...     ...       ...      ...   ...   ...   
1442        1    72    63  68.0      10        61     63.0     0     3   
1443        1    65    43  54.0      -3        45     49.0    11     0   
1444        1    47    35  41.0     -16        34     39.0    24     0   
1445        1    55    37  46.0     -11        35     42.0    19     0   
1446        1    61    42  52.0      -4        38     45.0    13     0   
1447        1    66    48  57.0       1        43     50.0     8     0   
1448        1    65    44  55.0      -1        35     46.0    10     0   
1449        1    62    43  53.0      -2        34     44.0    12     0   
1450        1    58    39  49.0      -6        39     44.0    16     0   
1451        1    59    36  48.0      -6        34     42.0    17     0   
1452        1    63    38  51.0      -3        42     47.0    14     0   
1453        1    67    56  62.0       9        58     60.0     3     0   
1454        1    65    52  59.0       6        57     58.0     6     0   
1455        1    55    50  53.0       0        51     51.0    12     0   
1456        1    56    50  53.0       1        49     51.0    12     0   
1457        1    62    48  55.0       3        46     50.0    10     0   
1458        1    50    36  43.0      -8        38     42.0    22     0   
1459        1    54    33  44.0      -7        38     43.0    21     0   
1460        1    65    46  56.0       6        44     49.0     9     0   
1461        1    53    45  49.0      -1        39     44.0    16     0   
1462        1    56    35  46.0      -4        38     43.0    19     0   
1463        1    59    36  48.0      -1        40     45.0    17     0   
1464      

In [62]:
# segregate Station_2
station2_df=combined.filter(regex='_2')
station2_df['Date']=combined['Date']
station2_df['DateStation']=station2_df['Date'].astype(str)+station2_df['Station_2'].astype(str)
print(station2_df.shape)
station2_df

(1472, 23)


C:\Users\User\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\User\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


Station_2  Tmax_2  Tmin_2  Tavg_2  Depart_2  DewPoint_2  WetBulb_2  \
0             2      84      52    68.0        14          51       57.0   
1             2      60      43    52.0        -3          42       47.0   
2             2      67      48    58.0         2          40       50.0   
3             2      78      51    64.0         4          42       50.0   
4             2      66      54    60.0         5          39       50.0   
5             2      68      52    60.0         4          30       46.0   
6             2      84      50    67.0        10          39       53.0   
7             2      80      60    70.0        12          57       63.0   
8             2      76      63    70.0        13          60       63.0   
9             2      83      59    71.0        14          52       61.0   
10            2      73      49    61.0         4          44       51.0   
11            2      65      47    56.0        -2          37       46.0   
12            2      69      44    57.0        -2          32       46.0   
13            2      90      54    72.0        15          45       58.0   
14            2      82      56    69.0        11          56       61.0   
15            2      64      45    55.0        -4          42       47.0   
16            2      61      44    53.0        -6          32       43.0   
17            2      70      43    57.0        -3          31       47.0   
18            2      80      56    68.0         8          37       53.0   
19            2      69      44    57.0        -5          41       45.0   
20            2      79      46    63.0         1          50       57.0   
21            2      87      60    74.0        10          44       58.0   
22            2      89      66    78.0        14          48       61.0   
23            2      88      66    77.0        14          53       63.0   
24            2      66      57    62.0         1          46       53.0   
25            2      70      50    60.0        -5          55       57.0   
26            2      75      59    67.0         4          49       57.0   
27            2      76      54    65.0         2          49       57.0   
28            2      88      65    77.0        10          59       66.0   
29            2      89      70    80.0        13          57       66.0   
...         ...     ...     ...     ...       ...         ...        ...   
1442          2      75      66    71.0        10          63       65.0   
1443          2      68      46    57.0        -3          47       52.0   
1444          2      49      37    43.0       -16          35       41.0   
1445          2      56      42    49.0       -11          36       44.0   
1446          2      63      48    56.0        -4          40       47.0   
1447          2      69      49    59.0         1          45       51.0   
1448          2      67      48    58.0        -1          36       48.0   
1449          2      64      50    57.0        -2          36       46.0   
1450          2      59      44    52.0        -6          40       46.0   
1451          2      60      40    50.0        -6          36       44.0   
1452          2      64      45    55.0        -3          44       50.0   
1453          2      71      59    65.0         9          61       62.0   
1454          2      67      54    61.0         6          59       60.0   
1455          2      58      53    56.0         0          53       54.0   
1456          2      58      53    56.0         1          52       53.0   
1457          2      63      49    56.0         3          48       52.0   
1458          2      52      39    46.0        -8          40       44.0   
1459          2      57      36    47.0        -7          40       44.0   
1460          2      66      50    58.0         6          46       51.0   
1461          2      54      49    52.0        -1          41       46.0   
1462          2      56      43    50.0        -4     

In [63]:
#Remove suffix
station2_df.columns = [col.replace('_2', '') for col in station2_df.columns]
station2_df

Station  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  Cool  \
0           2    84    52  68.0      14        51     57.0     0     3   
1           2    60    43  52.0      -3        42     47.0    13     0   
2           2    67    48  58.0       2        40     50.0     7     0   
3           2    78    51  64.0       4        42     50.0     7     0   
4           2    66    54  60.0       5        39     50.0     5     0   
5           2    68    52  60.0       4        30     46.0     5     0   
6           2    84    50  67.0      10        39     53.0     0     2   
7           2    80    60  70.0      12        57     63.0     0     5   
8           2    76    63  70.0      13        60     63.0     0     5   
9           2    83    59  71.0      14        52     61.0     0     6   
10          2    73    49  61.0       4        44     51.0     4     0   
11          2    65    47  56.0      -2        37     46.0     9     0   
12          2    69    44  57.0      -2        32     46.0     8     0   
13          2    90    54  72.0      15        45     58.0     0     7   
14          2    82    56  69.0      11        56     61.0     0     4   
15          2    64    45  55.0      -4        42     47.0    10     0   
16          2    61    44  53.0      -6        32     43.0    12     0   
17          2    70    43  57.0      -3        31     47.0     8     0   
18          2    80    56  68.0       8        37     53.0     0     3   
19          2    69    44  57.0      -5        41     45.0     8     0   
20          2    79    46  63.0       1        50     57.0     2     0   
21          2    87    60  74.0      10        44     58.0     0     9   
22          2    89    66  78.0      14        48     61.0     0    13   
23          2    88    66  77.0      14        53     63.0     0    12   
24          2    66    57  62.0       1        46     53.0     3     0   
25          2    70    50  60.0      -5        55     57.0     5     0   
26          2    75    59  67.0       4        49     57.0     0     2   
27          2    76    54  65.0       2        49     57.0     0     0   
28          2    88    65  77.0      10        59     66.0     0    12   
29          2    89    70  80.0      13        57     66.0     0    15   
...       ...   ...   ...   ...     ...       ...      ...   ...   ...   
1442        2    75    66  71.0      10        63     65.0     0     6   
1443        2    68    46  57.0      -3        47     52.0     8     0   
1444        2    49    37  43.0     -16        35     41.0    22     0   
1445        2    56    42  49.0     -11        36     44.0    16     0   
1446        2    63    48  56.0      -4        40     47.0     9     0   
1447        2    69    49  59.0       1        45     51.0     6     0   
1448        2    67    48  58.0      -1        36     48.0     7     0   
1449        2    64    50  57.0      -2        36     46.0     8     0   
1450        2    59    44  52.0      -6        40     46.0    13     0   
1451        2    60    40  50.0      -6        36     44.0    15     0   
1452        2    64    45  55.0      -3        44     50.0    10     0   
1453        2    71    59  65.0       9        61     62.0     0     0   
1454        2    67    54  61.0       6        59     60.0     4     0   
1455        2    58    53  56.0       0        53     54.0     9     0   
1456        2    58    53  56.0       1        52     53.0     9     0   
1457        2    63    49  56.0       3        48     52.0     9     0   
1458        2    52    39  46.0      -8        40     44.0    19     0   
1459        2    57    36  47.0      -7        40     44.0    18     0   
1460        2    66    50  58.0       6        46     51.0     7     0   
1461        2    54    49  52.0      -1        41     46.0    13     0   
1462        2    56    43  50.0      -4        38     44.0    15     0   
1463        2    62    41  52.0      -1        41     47.0    13     0   
1464      

In [66]:
weather_final=pd.concat([station1_df,station2_df])
print(weather_final.shape)
weather_final

(2944, 23)


Station  Tmax  Tmin  Tavg  Depart  DewPoint  WetBulb  Heat  Cool  \
0           1    83    50  67.0      14        51     56.0     0     2   
1           1    59    42  51.0      -3        42     47.0    14     0   
2           1    66    46  56.0       2        40     48.0     9     0   
3           1    66    49  58.0       4        41     50.0     7     0   
4           1    66    53  60.0       5        38     49.0     5     0   
5           1    68    49  59.0       4        30     46.0     6     0   
6           1    83    47  65.0      10        41     54.0     0     0   
7           1    82    54  68.0      12        58     62.0     0     3   
8           1    77    61  69.0      13        59     63.0     0     4   
9           1    84    56  70.0      14        52     60.0     0     5   
10          1    70    51  61.0       4        42     51.0     4     0   
11          1    64    46  55.0      -2        36     46.0    10     0   
12          1    69    43  56.0      -2        33     46.0     9     0   
13          1    90    56  73.0      15        47     59.0     0     8   
14          1    80    57  69.0      11        56     61.0     0     4   
15          1    66    44  55.0      -4        43     48.0    10     0   
16          1    62    43  53.0      -6        32     43.0    12     0   
17          1    71    42  57.0      -3        30     46.0     8     0   
18          1    80    56  68.0       8        39     53.0     0     3   
19          1    69    43  56.0      -5        40     44.0     9     0   
20          1    80    44  62.0       1        50     56.0     3     0   
21          1    87    55  71.0      10        46     58.0     0     6   
22          1    89    60  75.0      14        49     61.0     0    10   
23          1    88    63  76.0      14        54     63.0     0    11   
24          1    70    56  63.0       1        44     53.0     2     0   
25          1    65    49  57.0      -5        55     56.0     8     0   
26          1    75    57  66.0       4        47     56.0     0     1   
27          1    79    50  65.0       2        47     56.0     0     0   
28          1    88    60  74.0      10        58     65.0     0     9   
29          1    87    66  77.0      13        58     65.0     0    12   
...       ...   ...   ...   ...     ...       ...      ...   ...   ...   
1442        2    75    66  71.0      10        63     65.0     0     6   
1443        2    68    46  57.0      -3        47     52.0     8     0   
1444        2    49    37  43.0     -16        35     41.0    22     0   
1445        2    56    42  49.0     -11        36     44.0    16     0   
1446        2    63    48  56.0      -4        40     47.0     9     0   
1447        2    69    49  59.0       1        45     51.0     6     0   
1448        2    67    48  58.0      -1        36     48.0     7     0   
1449        2    64    50  57.0      -2        36     46.0     8     0   
1450        2    59    44  52.0      -6        40     46.0    13     0   
1451        2    60    40  50.0      -6        36     44.0    15     0   
1452        2    64    45  55.0      -3        44     50.0    10     0   
1453        2    71    59  65.0       9        61     62.0     0     0   
1454        2    67    54  61.0       6        59     60.0     4     0   
1455        2    58    53  56.0       0        53     54.0     9     0   
1456        2    58    53  56.0       1        52     53.0     9     0   
1457        2    63    49  56.0       3        48     52.0     9     0   
1458        2    52    39  46.0      -8        40     44.0    19     0   
1459        2    57    36  47.0      -7        40     44.0    18     0   
1460        2    66    50  58.0       6        46     51.0     7     0   
1461        2    54    49  52.0      -1        41     46.0    13     0   
1462        2    56    43  50.0      -4        38     44.0    15     0   
1463        2    62    41  52.0      -1        41     47.0    13     0   
1464      

In [68]:
weather_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2944 entries, 0 to 1471
Data columns (total 23 columns):
Station        2944 non-null int64
Tmax           2944 non-null int64
Tmin           2944 non-null int64
Tavg           2944 non-null float64
Depart         2944 non-null int64
DewPoint       2944 non-null int64
WetBulb        2944 non-null float64
Heat           2944 non-null int64
Cool           2944 non-null int64
Sunrise        2944 non-null int64
Sunset         2944 non-null int64
CodeSum        2944 non-null object
Depth          2944 non-null int64
Water1         2944 non-null object
SnowFall       2944 non-null object
PrecipTotal    2944 non-null object
StnPressure    2944 non-null float64
SeaLevel       2944 non-null float64
ResultSpeed    2944 non-null float64
ResultDir      2944 non-null int64
AvgSpeed       2944 non-null float64
Date           2944 non-null object
DateStation    2944 non-null object
dtypes: float64(6), int64(11), object(6)
memory usage: 552.0+ KB


In [ ]:
# datatype for PrecipTotal,Water1,SnowFall still as object

In [69]:
# export to csv
weather_final.to_csv('weather_final.csv',index=False)